targets/csharp/source/PlayFabSDK/source/Qos/PlayFabQosApi.cs

```diff
@@ -0,0 +1,163 @@
+﻿namespace PlayFab.QoS
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading.Tasks;
+    using EventsModels;
+    using Internal;
+    using MultiplayerModels;
+
+    public class PlayFabQosApi
+    {
+        private const int DefaultTimeoutMs = 250;
+        private readonly Dictionary<AzureRegion, string> _dataCenterMap = new Dictionary<AzureRegion, string>();
+        private readonly Action<string> _logAction;
+
+        public PlayFabQosApi(Action<string> logAction = null)
+        {
+            _logAction = logAction;
+        }
+
+        public async Task<QosResult> GetQosResultAsync(int timeoutMs = DefaultTimeoutMs)
+        {
+            QosResult result = await GetResultAsync(timeoutMs);
+            if (result.ErrorCode != (int)QosErrorCode.Success)
+            {
+                return result;
+            }
+
+            Task.Factory.StartNew(() => SendResultsToPlayFab(result));
+            return result;
+        }
+
+        private async Task<QosResult> GetResultAsync(int timeoutMs)
+        {
+            var result = new QosResult();
+
+            if (!PlayFabClientAPI.IsClientLoggedIn())
+            {
+                LogQos("Client is not logged in");
+                result.ErrorCode = (int)QosErrorCode.NotLoggedIn;
+                return result;
+            }
+
+            // get datacenter map (call thunderhead)
+            await GetQoSServerList();
+
+            int serverCount = _dataCenterMap.Count;
+            if (serverCount <= 0)
+            {
+                result.ErrorCode = (int)QosErrorCode.FailedToRetrieveServerList;
+                return result;
+            }
+
+            // ping servers
+            result.RegionResults = await GetSortedRegionLatencies(timeoutMs);
+            result.ErrorCode = (int)QosErrorCode.Success;
+            return result;
+        }
+
+        private async Task GetQoSServerList()
+        {
+            if (_dataCenterMap.Count > 0)
+            {
+                // If the dataCenterMap is already initialized, return
+                return;
+            }
+
+            var request = new ListQosServersRequest();
+            PlayFabResult<ListQosServersResponse> response = await PlayFabMultiplayerAPI.ListQosServersAsync(request);
+            
+            if (response == null || response.Error != null)
+            {
+                LogQos("Could not get the server list from thunderhead\n Error : " + response?.Error.GenerateErrorReport());
```

Comment:
You should do a global replace for:
"response?.Error."
to
"response?.Error?."